Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -143,3 +146,77 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
   (Relation.ReflTransGen.cases_tail huw).casesOn (Or.inl ∘ Eq.symm) Or.inr
 
 end ContextFreeGrammar
+
+/-- Context-free languages are defined by context-free grammars. -/
+def Language.IsContextFree (L : Language T) : Prop :=
+  ∃ g : ContextFreeGrammar.{uT} T, g.language = L
+
+/-- `L.reverse` is a language that contains exactly all words from `L` backwards. -/
+def Language.reverse (L : Language T) : Language T := { w : List T | w.reverse ∈ L }
+
+private def reverseRule {N : Type uN} (r : ContextFreeRule T N) : ContextFreeRule T N :=
+  ⟨r.input, r.output.reverse⟩
+
+private def reverseGrammar (g : ContextFreeGrammar T) : ContextFreeGrammar T :=
+  ⟨g.NT, g.initial, g.rules.map reverseRule⟩
+
+private lemma reverseRule_reverseRule {N : Type uN} (r : ContextFreeRule T N) :
+    reverseRule (reverseRule r) = r := by
+  simp [reverseRule]
+
+private lemma dual_reverseRule {N : Type uN} : reverseRule ∘ @reverseRule T N = id := by
+  ext
+  apply reverseRule_reverseRule
+
+private lemma reverseGrammar_reverseGrammar (g : ContextFreeGrammar T) :
+    reverseGrammar (reverseGrammar g) = g := by
+  cases g with
+  | mk => simp [reverseGrammar, dual_reverseRule, List.map_map, List.map_id]
+
+private lemma derives_reverse {g : ContextFreeGrammar T} {s : List (Symbol T g.NT)}
```

Comment:
I'm not sure all of these things should be private, making things private generally just makes it harder for other people to prove things about them

Reply:
I guess the question is if there are any possible other results whose proof would make use of `reverseGrammar`?

Reply:
I don't think so.